src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -147,10 +147,13 @@ static SemanticTokensHelpers()
 
             // If the full compilation is not yet available, we'll try getting a partial one. It may contain inaccurate
             // results but will speed up how quickly we can respond to the client's request.
-            var frozenDocument = document.WithFrozenPartialSemantics(cancellationToken);
-            var semanticModel = await frozenDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
-            var isFinalized = document.Project.TryGetCompilation(out var compilation) && compilation == semanticModel.Compilation;
-            document = frozenDocument;
+            document = document.WithFrozenPartialSemantics(cancellationToken);
+
+            // Razor uses isFinalized to determine whether to cache tokens. IsFullyLoaded may not contain the fully finalized
+            // tokens, but we should be able to remove it altogether once Roslyn implements workspace/semanticTokens/refresh:
+            // https://github.com/dotnet/roslyn/issues/60441
+            var workspaceStatusService = document.Project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+            var isFinalized = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I think we need to only set isFinalized when both IsFullyLoaded is true _and_ the getcompilation task for the fully loaded solution is done (I think)

Otherwise the project might be fully loaded, but the compilation not complete so we give razor wrong tokens and cache wrong tokens